File: vbnet/Tennis/TennisGame3.vb
Public Class TennisGame3
    Implements ITennisGame

    Private p2 As Integer
    Private p1 As Integer
    Private p1N As String
    Private p2N As String

    Public Sub New(ByVal player1Name As String, ByVal player2Name As String)
        Me.p1N = player1Name
        Me.p2N = player2Name
    End Sub

    Public Function GetScore() As String Implements ITennisGame.GetScore
        Dim s As String

        If (p1 < 4 AndAlso p2 < 4) AndAlso (p1 + p2 < 6) Then
            Dim p As String() = {"Love", "Fifteen", "Thirty", "Forty"}
            s = p(p1)
            Return If((p1 = p2), s & "-All", s & "-" & p(p2))
        Else
            If p1 = p2 Then Return "Deuce"
            s = If(p1 > p2, p1N, p2N)
            Return If(((p1 - p2) * (p1 - p2) = 1), "Advantage " & s, "Win for " & s)
        End If
    End Function

    Public Sub WonPoint(ByVal playerName As String) Implements ITennisGame.WonPoint
        If playerName = "player1" Then
            Me.p1 += 1
        Else
            Me.p2 += 1
        End If
    End Sub
End Class

File: csharp/Tennis/TennisGame3.cs
namespace Tennis
{
    public class TennisGame3 : ITennisGame
    {
        private int player2Points;
        private int player1Points;
        private string player1Name;
        private string player2Name;
        public static readonly string[] POINTS_NAMES = new[] { "Love", "Fifteen", "Thirty", "Forty" };

        public TennisGame3(string player1Name, string player2Name)
        {
            this.player1Name = player1Name;
            this.player2Name = player2Name;
        }

        public string GetScore()
        {
            var isNotYetEndgame = (player1Points < 4 && player2Points < 4) && (player1Points + player2Points < 6);
            if (isNotYetEndgame)
            {
                var score = POINTS_NAMES[player1Points];
                if (player1Points == player2Points)
                    return score + "-All";
                else
                    return score + "-" + POINTS_NAMES[player2Points];
            }
            else
            {
                if (player1Points == player2Points)
                    return "Deuce";
                var leader = player1Points > player2Points ? player1Name : player2Name;
                var pointsDifferenceIsOne = (player1Points - player2Points) * (player1Points - player2Points) == 1;
                if (pointsDifferenceIsOne)
                    return "Advantage " + leader;
                else
                    return "Win for " + leader;
            }
        }

        public void WonPoint(string playerName)
        {
            if (playerName == "player1")
                this.player1Points += 1;
            else
                this.player2Points += 1;
        }

    }
}

